js/share_test.go

```diff
@@ -0,0 +1,173 @@
+ *
+ */
+
+package js
+
+import (
+	"context"
+	"io/ioutil"
+	"testing"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/testutils"
+	"github.com/loadimpact/k6/stats"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInitContextNewSharedArray(t *testing.T) {
+	data := `'use strict';
+var SharedArray = require("k6/data").SharedArray;
+function generateArray() {
+    console.log("once");
+    var n = 50;
+    var arr = new Array(n);
+    for (var i = 0 ; i <n; i++) {
+        arr[i] = {value: "something" +i};
+    }
+    return arr;
+}
+
+var s = new SharedArray("something", generateArray);
+
+try {
+	var p = new SharedArray("wat", function() {return "whatever"});
+	throw "the previous line should've errored as we returned a string not array";
+} catch (e) {
+	if (!e.toString().includes("only arrays can be made into SharedArray")) {
+		throw "wrong error " + e.toString();
+	}
+}
+
+try {
+	var p = new SharedArray("wat2", function() {return [{s: function() {return "whatever"}}]});
+	// throw "the previous line should've errored as function was stringified";
+	// unfortunately JSON.stringify is defined that it will acctually just remove stuff (or make them null)
+	// that can't be JSONified like functions
+	if (p[0].s != undefined) {
+		throw "only arrays can be made into SharedArray";
+	}
+} catch (e) {
+	if (!e.toString().includes("only arrays can be made into SharedArray")) {
+		throw "wrong error " + e.toString();
+	}
+}
+
+
+exports.default = function() {
+	if (s[2].value !== "something2") {
+		throw new Error("bad s[2]="+s[2].value);
+	}
+	if (s.length != 50) {
+		throw new Error("bad length " +_s.length);
+	}
+
+	var i = 0;
+	for (var v of s) {
+		if (v.value !== "something"+i) {
+			throw new Error("bad v.value="+v.value+" for i="+i);
+		}
+		i++;
+
+		try {
+			v.data = "help";
+			throw "the previous line should've errored v.data = 'help'";
+		} catch(e) {
+			if (!e.toString().includes("TypeError: Cannot add property data, object is not extensible")) {
+				throw "wrong error " + e.toString();
+			}
+		}
+	}
+
+	try {
+		s[2].data = "help";
+		throw "the previous line should've errored s[2].data = 'help'";
+	} catch(e) {
+		if (!e.toString().includes("TypeError: Cannot add property data, object is not extensible")) {
+			throw "wrong error " + e.toString();
+		}
+	}
+
+	if (s.something != undefined) {
+		throw "s.something should've been undefined but was " + s.something;
+	}
+
+	try {
+		s.something = 21
+		throw "the previous line should've errored s.something = 21";
+	} catch(e) {
+		if (!e.toString().includes("Host object field something cannot be made configurable")) {
+			throw "wrong error " + e.toString();
+		}
+	}
+
+	try {
+		s[1]= "21";
+
+		throw "the previous line should've errored";
+	} catch(e) {
+		if (!e.toString().includes("Host object field 1 cannot be made configurable")) {
+			throw "wrong error " + e.toString();
```

Comment:
I would prefer to see separate Go testcases for each scenario, rather than this large JS script. It would be easier to run them individually and also to modify or add new ones.

Reply:
Hm ... maybe, it used to be fairly short. On the other hand that will increase the amount of time the test takes(goja initialization isn't free even without the babel and corejs), while changes to this test seem ... unlikely unless we add functionality and I don't think we will be adding all that much as I doubt a lot of users will want additional Array like functionality.  


I am even starting to question if enabling for-of was worth it ...

Reply:
I would still prefer legibility/usability over a slightly longer run time, but it's fine if you don't think it's worth it.

I saw your comment about the 100-200x performance penalty with the iterator, and now I'm not so sure if we should keep it either 😄 I would assume access by index will be more common, and looping by index would work as well, but some users might prefer the `for-of` notation and consider the performance loss acceptable(?). We need documentation to explain this either way, so I'm leaning towards leaving it in as is.